### PR TITLE
Fix backend tests for group discounts, new cart API, and more

### PR DESCRIPTION
**Pull request by aviupadhyayula**
Changelog: 
1. Refactors total price summation to an internal helper function. Adds tests for it. (Previous code was failing but wasn't caught, since we mock `initiate_checkout` and `complete_checkout`. Breaking this logic into its own, testable function will hopefully catch future mistakes.)
2. Adds tests for creating tickets with group discounts.
3. Adds tests for summing a cart total with group discounts applied. 
4. Only populates `sold_out` if there are tickets that were not replaced.
5. Aligns tests with changes introduced in ^ and #670.

### CODE REVIEW

File: backend/clubs/views.py [hunk 1]
```diff
@@ -4945,21 +4969,7 @@ def initiate_checkout(self, request, *args, **kwargs):
...
-
-        cart_total = sum(
-            (
-                ticket.price * (1 - ticket.group_discount)
-                if ticket.group_size
-                and ticket_type_counts[ticket.type] >= ticket.group_size
-                else ticket.price
-            )
-            for ticket in tickets
-        )
```
Comment (aviupadhyayula) [inline backend/clubs/views.py:4962]: Was [previously broken](https://stackoverflow.com/questions/45768190/typeerror-manyrelatedmanager-object-is-not-iterable).

File: backend/clubs/views.py [hunk 2]
```diff
@@ -4782,6 +4783,29 @@ class TicketViewSet(viewsets.ModelViewSet):
     http_method_names = ["get", "post"]
     lookup_field = "id"
 
+    def _calculate_cart_total(self, cart) -> float:
```
Comment (rohangpta) [inline backend/clubs/views.py:4786]: Could we make this a free function? `self` isn't used in the body anywhere.

File: backend/tests/clubs/test_ticketing.py [hunk 3]
```diff
@@ -525,7 +554,48 @@ def test_get_cart(self):
...
+                type="group",
+                event=self.event1,
+                price=10.0,
+                group_size=2,
+                group_discount=0.2,
+            )
+            for _ in range(10)
+        ]
+
+        # Add enough to activate group discount
```
Comment (rohangpta) [inline backend/tests/clubs/test_ticketing.py:588]: Could you also assert the correct cart total before hitting the threshold for group discount?

File: backend/tests/clubs/test_ticketing.py [hunk 4]
```diff
@@ -525,7 +554,48 @@ def test_get_cart(self):
         self.assertEqual(len(data["tickets"]), 5, data)
         for t1, t2 in zip(data["tickets"], tickets_to_add):
             self.assertEqual(t1["id"], str(t2.id))
-        self.assertEqual(data["sold_out"], 0, data)
+        num_sold_out = sum(item["count"] for item in data["sold_out"])
```
Comment (rohangpta) [inline backend/tests/clubs/test_ticketing.py:557]: Please assert on all keys and values of `data["sold_out"]`, below as well.

File: backend/tests/clubs/test_ticketing.py [hunk 5]
```diff
@@ -741,9 +814,10 @@ def test_complete_checkout(self):
             self.assertEqual(held_tickets.count(), 0, held_tickets)
 
             # Transaction record created
-            TicketTransactionRecord.objects.filter(
+            record_exists = TicketTransactionRecord.objects.filter(
                 reconciliation_id=MockPaymentResponse().reconciliation_id
             ).exists()
+            self.assertTrue(record_exists)
```
Comment (rohangpta) [inline backend/tests/clubs/test_ticketing.py:820]: Nice catch!

**Review by rohangpta — COMMENTED** [review]
Nice work! Looks like these APIs have moved but the tests haven't quite caught up - I've added a few comments around that.

File: backend/clubs/views.py [hunk 6]
```diff
@@ -4802,7 +4826,7 @@ def cart(self, request, *args, **kwargs):
                                     allOf:
                                         - $ref: "#/components/schemas/Ticket"
                                 sold_out:
-                                    type: list
+                                    type: array
```
Comment (joyliu-q) [inline backend/clubs/views.py:4829]: Oops thanks for the catch haha

**Review by joyliu-q — APPROVED** [review]
Looks great, thanks for the catches!

File: backend/clubs/views.py (hunk 2 above)
Comment (aviupadhyayula) [reply backend/clubs/views.py]: Made it a static method.

File: backend/clubs/views.py [hunk 7]
```diff
@@ -4802,25 +4827,14 @@ def cart(self, request, *args, **kwargs):
...
-                                                        type: string
+                                                type: integer
                                             type:
                                                 type: string
-                                            owner:
-                                                type: string
-                                            price:
-                                                type: number
                                             count:
                                                 type: integer
```
Comment (aviupadhyayula) [inline backend/clubs/views.py:4839]: @joyliu-q looks like you're dealing with the annotated queryset (see L4865), so these fields won't exist. Unsure if this is the behavior you intended.

Comment (julianweng) [reply backend/clubs/views.py:4839]: Fixing this shortly

Comment (aviupadhyayula) [reply backend/clubs/views.py]: Thanks! Looks like there were some merge conflicts as a result of your changes. Resolved conflicts in [a9054ea](https://github.com/pennlabs/penn-clubs/pull/675/commits/a9054eab62c8e7c2556059ba5b9796fe7bf68d9a) and modified tests in [63ff452](https://github.com/pennlabs/penn-clubs/pull/675/commits/63ff45235b498aaa77026da50e7dd3fa0eb99b04).

**Pull request merged**